modules/navigation/3d/nav_mesh_queries_3d.cpp

```diff
@@ -681,57 +691,67 @@ void NavMeshQueries3D::_path_corridor_post_process_edgecentered(NavMeshPathQuery
 
-	_query_task_push_back_point_with_metadata(p_query_task, p_end_point, p_end_polygon);
+	_query_task_push_back_point_with_metadata(p_query_task, end_point, p_query_task.end_poly);
 
 	// Add mid points.
-	int np_id = p_least_cost_id;
+	int np_id = p_query_task.least_cost_id;
 	while (np_id != -1 && p_path_corridor[np_id].back_navigation_poly_id != -1) {
 		_query_task_push_back_point_with_metadata(p_query_task, p_path_corridor[np_id].entry, p_path_corridor[np_id].poly);
 
 		np_id = p_path_corridor[np_id].back_navigation_poly_id;
 	}
 
-	_query_task_push_back_point_with_metadata(p_query_task, p_begin_point, p_begin_poly);
+	_query_task_push_back_point_with_metadata(p_query_task, begin_point, p_query_task.begin_poly);
 }
 
-void NavMeshQueries3D::_query_task_find_start_end_positions(NavMeshPathQueryTask3D &p_query_task, const LocalVector<gd::Polygon> &p_polygons, const gd::Polygon **r_begin_poly, Vector3 &r_begin_point, const gd::Polygon **r_end_poly, Vector3 &r_end_point) {
+void NavMeshQueries3D::_query_task_find_start_end_positions(NavMeshPathQueryTask3D &p_query_task, const LocalVector<gd::Polygon> &p_polygons) {
+	// Find begin polyon and begin position closest to start position and
+	// end polyon and end position closest to target position on the map.
 	real_t begin_d = FLT_MAX;
 	real_t end_d = FLT_MAX;
 
+	Vector3 begin_point;
+	Vector3 end_point;
+
 	// Find the initial poly and the end poly on this map.
-	for (const gd::Polygon &p : p_polygons) {
+	for (const gd::Polygon &polygon : p_polygons) {
 		// Only consider the polygon if it in a region with compatible layers.
-		if ((p_query_task.navigation_layers & p.owner->get_navigation_layers()) == 0) {
+		if ((p_query_task.navigation_layers & polygon.owner->navigation_layers) == 0) {
 			continue;
 		}
 
 		// For each face check the distance between the origin/destination.
-		for (size_t point_id = 2; point_id < p.points.size(); point_id++) {
-			const Face3 face(p.points[0].pos, p.points[point_id - 1].pos, p.points[point_id].pos);
+		for (size_t point_id = 2; point_id < polygon.points.size(); point_id++) {
+			const Face3 face(polygon.points[0].pos, polygon.points[point_id - 1].pos, polygon.points[point_id].pos);
 
 			Vector3 point = face.get_closest_point_to(p_query_task.start_position);
 			real_t distance_to_point = point.distance_to(p_query_task.start_position);
 			if (distance_to_point < begin_d) {
 				begin_d = distance_to_point;
-				*r_begin_poly = &p;
-				r_begin_point = point;
+				p_query_task.begin_poly = &polygon;
+				begin_point = point;
 			}
 
 			point = face.get_closest_point_to(p_query_task.target_position);
 			distance_to_point = point.distance_to(p_query_task.target_position);
 			if (distance_to_point < end_d) {
 				end_d = distance_to_point;
-				*r_end_poly = &p;
-				r_end_point = point;
+				p_query_task.end_poly = &polygon;
+				end_point = point;
 			}
 		}
 	}
+
+	p_query_task.begin_point = begin_point;
```

Comment:
Yes but in that case the polygon pointer will be empty so I did not add another check for the positions. If the polygon pointer is empty after returning from this function it will make the query finish and return empty.